fix(robot-server): get versions from VERSION.json

sfoster1 opened the pull request:
In d7a9200b102d4e0ffbe06c483d2fdb1fb1765016 we switched to generating the version of the opentrons package from git, including feeding its __version__ by using importlib metadata instead of ingesting package.json. The problem is that __version__ now has a python-style semver string (e.g. 0.1.0a0) instead of a rest-of-the-world semver string (e.g. 0.1.0-alpha.0), which node semver doesn't like, which means that the app was rejecting the version the robot server reported in /health and falling back to the update server. This means things continued to work perfectly fine on OT-2, but on the flex a separate issue meant the update server wasn't reporting a version for the opentrons package, so versions broke.

This change uses the VERSION.json file to get the "real" semver string for the opentrons package instead of the importlib metadata one. It also uses that file to load the system version rather than a config flag.

There are fallbacks to the original versions of this implementation for both, which will be used in dev servers where VERSION.json is not present.

Closes RQA-577

## Risks
- Well, we have to hit the version file now, but this is a run-once dependency so it should be fine

## Review Requests
- This is robot server code, is this okay to be doing?

## Testing
- [x] Push to a flex and check that we get the right version
- [x] push to an OT-2 and check that we get the right version

sfoster1 committed:
fix(robot-server): get versions from VERSION.json

In d7a9200b102d4e0ffbe06c483d2fdb1fb1765016 we switched to generating
the version of the opentrons package from git, including feeding its
__version__ by using importlib metadata instead of ingesting
package.json. The problem is that __version__ now has a python-style
semver string (e.g. 0.1.0a0) instead of a rest-of-the-world semver
string (e.g. 0.1.0-alpha.0), which node semver doesn't like, which means
that the app was rejecting the version the robot server reported in
/health and falling back to the update server. This means things
continued to work perfectly fine on OT-2, but on the flex a separate
issue meant the update server wasn't reporting a version for the
opentrons package, so versions broke.

This change uses the VERSION.json file to get the "real" semver string
for the opentrons package instead of the importlib metadata one. It also
uses that file to load the system version rather than a config flag.

There are fallbacks to the original versions of this implementation for
both, which will be used in dev servers where VERSION.json is not
present.

Closes RQA-577

SyntaxColoring commented on robot-server/robot_server/health/router.py:
```diff
@@ -9,8 +13,78 @@
+    except FileNotFoundError:
+        _log.warning(f"{VERSION_PATH} does not exist - is this a dev server?")
+        return {}
+    except OSError as ose:
+        _log.warning(
+            f"Could not open {VERSION_PATH}: {ose.errno}: {ose.strerror} - is this a dev server?"
+        )
+        return {}
+    except json.JSONDecodeError as jde:
+        _log.error(
+            f"Could not parse {VERSION_PATH}: {jde.msg} at line {jde.lineno} col {jde.colno}"
+        )
+        return {}
```
If it makes it easier, instead of constructing a message out of `.errno` and such, I think you can do `exc_info=True` to get the logger to include the full details of the current exception.

```suggestion
    except OSError as ose:
        _log.warning(
            f"Could not open {VERSION_PATH} - is this a dev server?",
            exc_info=True,
        )
        return {}
    except json.JSONDecodeError as jde:
        _log.error(
            f"Could not parse {VERSION_PATH}",
            exc_info=True,
        )
        return {}
```

sfoster1 replied:
> It's surprising to me that the version string is semantically meaningful enough to cause any kind of breakage if it says `a` instead of `alpha`. I'll have to catch up on the details of the bug, but it seems like we should either not have the app assume any specific format, or make it really impossible for the server to return anything outside that specific format.

Yeah this is because [node-semver](https://github.com/npm/node-semver) doesn't like that format. Maybe there's a way to get it to like that format, and probably we should also do that, but this makes the robot server more compliant with the update server anyway.

sfoster1 replied:
Yeah, I don't really like that because it makes it a little clearer what you do and don't get with the explicit formatting. Plus it makes sure we don't get a bunch of big tracebacks when we already know where the problem is.

SyntaxColoring commented on robot-server/robot_server/health/router.py:
```diff
@@ -9,8 +13,78 @@
+
+    def _api_version_or_fallback() -> str:
+        if "opentrons_api_version" in version_file:
+            return version_file["opentrons_api_version"]
```
Do we need to care about the case where `version_file` is present, and contains valid JSON, but is missing the `opentrons_api_version` key?

SyntaxColoring replied:
Never mind, I'm dumb, you're doing this on line 65.

sfoster1 replied:
yeah, that was actually the case on flices before https://github.com/Opentrons/oe-core/pull/72 . I mean, it's a bug, but it's nice to be tolerant to it since we do have a fallback.

SyntaxColoring commented on robot-server/robot_server/health/router.py:
```diff
@@ -9,8 +13,78 @@
+_log = logging.getLogger(__name__)
 
 LOG_PATHS = ["/logs/serial.log", "/logs/api.log", "/logs/server.log"]
+VERSION_PATH = "/etc/VERSION.json"
```
Do you know if this file will be present if a customer manually rewrites the SD card with a fresh OS image, and doesn't do any of the factory provisioning stuff?

sfoster1 replied:
It will be, it's part of the (buildroot, openembedded) build.

SyntaxColoring commented on robot-server/tests/conftest.py:
```diff
@@ -92,6 +93,14 @@ def hardware() -> MagicMock:
     return MagicMock(spec=API)
 
 
+@pytest.fixture
+def versions() -> MagicMock:
+    m = MagicMock(spec=ComponentVersions)
+    m.api_version = "someTestApiVersion"
+    m.system_version = "someTestSystemVersion"
+    return m
```
It looks like this could be a real `ComponentVersions` instead of a mock?

```suggestion
@pytest.fixture
def versions() -> ComponentVersions:
    return ComponentVersions(
        api_version="someTestApiVersion",
        system_version="someTestSystemVersion",
    )
```

sfoster1 commented:
With this PR, on a flex:
```
curl 10.13.11.47:31950/health -H 'opentrons-version: *'
{"name":"FlexDVT","robot_model":"OT-3 Standard","api_version":"0.1.0-alpha.0","fw_version":"unknown","board_revision":"UNKNOWN","logs":["/logs/serial.log","/logs/api.log","/logs/server.log"],"system_version":"v0.1.0-1-gaca02cb","maximum_protocol_api_version":[2,14],"minimum_protocol_api_version":[2,0],"links":{"apiLog":"/logs/api.log","serialLog":"/logs/serial.log","serverLog":"/logs/server.log","apiSpec":"/openapi.json","systemTime":"/system/time"}}```

SyntaxColoring commented on robot-server/robot_server/health/router.py:
```diff
@@ -9,8 +13,78 @@
+        return {}
+    except OSError as ose:
+        _log.warning(
+            f"Could not open {VERSION_PATH}: {ose.errno}: {ose.strerror} - is this a dev server?"
```
What are the implications if this *is* a dev server?

Your PR description says how something broke because the Flex's robot-server was reporting a Python-style version string and its update-server wasn't reporting anything. Which sounds exactly like what will happen with a robot-server dev server.

Was it a dangerous kind of breakage, or was it just that a version label was displaying blank?

sfoster1 replied:
The version label displays blank, and the update flow fails because of it. It's fine for a dev server.

sfoster1 commented:
With this PR, on an OT-2:
```
curl 10.13.11.66:31950/health -H 'opentrons-version: *'
{"name":"Oooootie","robot_model":"OT-2 Standard","api_version":"6.3.0-alpha.1","fw_version":"v1.1.0-25e5cea","board_revision":"2.1","logs":["/logs/serial.log","/logs/api.log","/logs/server.log"],"system_version":"v1.14.4","maximum_protocol_api_version":[2,14],"minimum_protocol_api_version":[2,0],"links":{"apiLog":"/logs/api.log","serialLog":"/logs/serial.log","serverLog":"/logs/server.log","apiSpec":"/openapi.json","systemTime":"/system/time"}}```

SyntaxColoring commented on robot-server/robot_server/health/router.py:
```diff
@@ -9,8 +13,78 @@
+    return __version__
+
+
+async def get_versions() -> ComponentVersions:
```
I don't think it makes any practical difference, but your PR description mentioned this was a "run-once dependency." Was this meant to have `@call_once` or something to implement that?

sfoster1 replied:
Yes... is that not default behavior?

sfoster1 replied:
no it is definitely not. I guess it doesn't really matter a ton, but it's a good thing to decorate with call_once yeah

SyntaxColoring reviewed (approved):
Looks like it'll do what it says on the tin.

It's surprising to me that the version string is semantically meaningful enough to cause any kind of breakage if it says `a` instead of `alpha`. I'll have to catch up on the details of the bug, but it seems like we should either not have the app assume any specific format, or make it really impossible for the server to return anything outside that specific format.

sfoster1 committed:
Update robot-server/tests/conftest.py

Co-authored-by: Max Marrone <max@opentrons.com>

sfoster1 committed:
make versions call_once

Since it reads the version file, don't want to do that every request.

Also make that versions mock actually an async magicmock, because that
way test functions can request it and then change its value.

sfoster1 committed:
not quite sure why it didn't like this
